### PR TITLE
Updated esi security recommendation document.

### PR DESCRIPTION
**Pull request by mnbatra**
Added iDRAC reset and import/export information to the esi security recommendation document.

### CODE REVIEW

File: docs/source/usage/security_recommendations.rst [hunk 1]
```diff
@@ -112,5 +112,203 @@ Note:
...
+  python3 ImportSystemConfigurationLocalREDFISH.py -ip '10.1.10.3' -u 'new-user' -p $ADMIN_PASSWORD -t IDRAC
+
+
+Note:
+ * Redfish does not support exporting the iDRAC license iDRAC 7/8, RACADM command can be used instead.
+      * Usage
+           - racadm license help import
+           - racadm license help export
+
+Import or export iDRAC configuration with manual cleaning (Owner):
```
[the remark this answers is not in the document]
Comment (tzumainn) [reply docs/source/usage/security_recommendations.rst:256]: Also, I'd remove (Owner); this is for admins and owners.

**Review by tzumainn — CHANGES_REQUESTED** [review]
Hi! This is a good start. I think there may be *too much* detail to be honest; we want to give users a quick stripped-down version of what they need to know to accomplish their task, with outward links if they want to know more. I've marked some places where I think that detail can be reduced, but keep that idea in mind.

File: docs/source/usage/security_recommendations.rst [hunk 2]
```diff
@@ -112,5 +112,203 @@ Note:
...
+.. prompt::
+
+  [DEFAULT]
+  enabled_hardware_types=idrac
+  enabled_management_interfaces=idrac-wsman
+  enabled_power_interfaces=idrac-wsman
+  enabled_vendor_interface=idrac-wsman
+  enabled_bios_interface=idrac-wsman
+
+* To enable all optional features (BIOS, inspection, RAID, and vendor passthru) using Redfish where it is supported and WSMAN where not, use the following configuration:
```
Comment (DanNiESh) [inline docs/source/usage/security_recommendations.rst:196]: Maybe also mention that ironic_conductor service needs a restart to take effect.

File: docs/source/usage/security_recommendations.rst [hunk 3]
```diff
@@ -112,5 +112,203 @@ Note:
...
+
+Import or export iDRAC configuration with idrac-redfish:
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+A sample script using `Redfish API`_ to export system configurations.
+
+* Export configuration:
+
+.. prompt:: bash $
+  python3 ExportSystemConfigurationLocalREDFISH.py -ip '10.1.10.3' -u 'new-user' -p $ADMIN_PASSWORD -t IDRAC
```
[the remark this answers is not in the document]
Comment (DanNiESh) [reply docs/source/usage/security_recommendations.rst:241]: I think you need an empty line between ".. prompt:: bash $" and python command.

File: docs/source/usage/security_recommendations.rst [hunk 4]
```diff
@@ -112,5 +112,203 @@ Note:
...
+This happens because with UEFI boot mode iDRAC uses BIOS settings to manage the PXE configuration. For now, the only solution is for the operator to visit the data center and manually restore the PXE settings of the server to enable PXE boot again.
+
+The **preferred way** of restoring iDRAC to a known good state is to export a system configuration mold (JSON settings file) before leasing a node and import the configuration mold to restore the iDRAC settings after the lease ends. Two ways to import/export iDRAC settings:
+
+* Import/Export iDRAC configuration with REST APIs using Redfish.
+* Import/Export iDRAC configuration OpenStack Ironic Bare Metal Cleaning service.
+
+Configuration molds currently support only the OEM settings section that uses the iDRAC Server Configuration Profile (SCP) and can be edited as necessary if it complies with the SCP. The mold can be stored in either a Swift container or an HTTP server that has support for 'PUT' requests. More information about SCP: `SCP_Reference_Guide`_.
```
[the remark this answers is not in the document]
Comment (mnbatra) [reply docs/source/usage/security_recommendations.rst:126]: Removed the unnecessary bits

File: docs/source/usage/security_recommendations.rst [hunk 5]
```diff
@@ -112,5 +112,203 @@ Note:
...
+
+.. prompt:: bash $
+
+  sudo pip install 'python-dracclient>=3.1.0' 'sushy>=2.0.0'
+
+The iDRAC driver supports WSMAN for the bios, inspect, management, power, raid, and vendor interfaces. In addition, it supports Redfish for the bios, inspect, management, power, and raid interfaces. The iDRAC driver allows you to mix and match WSMAN and Redfish interfaces.
+
+* Enable the iDRAC hardware type with the required interfaces idrac-wsman
+
+.. prompt::
```
[the remark this answers is not in the document]
Comment (mnbatra) [reply docs/source/usage/security_recommendations.rst:187]: Linked to the official documentation

File: docs/source/usage/security_recommendations.rst (hunk 2 above)
Comment (mnbatra) [reply docs/source/usage/security_recommendations.rst:196]: Done!

File: docs/source/usage/security_recommendations.rst [hunk 6]
```diff
@@ -112,5 +112,203 @@ Note:
...
+  enabled_hardware_types=idrac
+  enabled_bios_interfaces=idrac-redfish
+  enabled_inspect_interfaces=idrac-redfish
+  enabled_management_interfaces=idrac-redfish
+  enabled_power_interfaces=idrac-redfish
+  enabled_raid_interfaces=idrac-redfish
+  enabled_vendor_interfaces=idrac-redfish
+
+
+Configure the Node
```
[the remark this answers is not in the document]
Comment (mnbatra) [reply docs/source/usage/security_recommendations.rst:210]: Moved to a sub section under pre-requisites

File: docs/source/usage/security_recommendations.rst [hunk 7]
```diff
@@ -112,5 +112,203 @@ Note:
...
+    --inspect-interface idrac-redfish \
+    --management-interface idrac-redfish \
+    --power-interface idrac-redfish
+    <node>
+
+
+Import or export iDRAC configuration with idrac-redfish:
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+A sample script using `Redfish API`_ to export system configurations.
```
[the remark this answers is not in the document]
Comment (mnbatra) [reply docs/source/usage/security_recommendations.rst:236]: I could not find direct example, but this link mentions every module's description https://github.com/dell/iDRAC-Redfish-Scripting/blob/master/iDRAC%20Python%20Redfish%20Module/README.txt

[the remark this answers is not in the document]
Comment (mnbatra) [reply docs/source/usage/security_recommendations.rst:236]: The scripts mentioned in this example were tested by @DanNiESh

File: docs/source/usage/security_recommendations.rst (hunk 3 above)
[the remark this answers is not in the document]
Comment (mnbatra) [reply docs/source/usage/security_recommendations.rst:241]: added \n

File: docs/source/usage/security_recommendations.rst (hunk 1 above)
[the remark this answers is not in the document]
Comment (mnbatra) [reply docs/source/usage/security_recommendations.rst:256]: Done

File: docs/source/usage/security_recommendations.rst [hunk 8]
```diff
@@ -112,5 +112,203 @@ Note:
...
+
+Import or export iDRAC configuration with manual cleaning (Owner):
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The clean and deploy steps provided in this section allow an owner to configure the system and collect the system inventory using configuration mold files. To export or import the configuration of current system or replicate the system configuration to that of a similar system.
+
+
+**Steps**:
+
+1. Configure a golden, or one to many, system.
```
[the remark this answers is not in the document]
Comment (mnbatra) [reply docs/source/usage/security_recommendations.rst:264]: Ironic document  mentions that we can export configuration files from one system, remove connection settings and other  unnecessary information and import it in a different system, thus one to many i think

> To replicate the system configuration to that of a similar system, perform the following steps:
> 
> Configure a golden, or one to many, system.
> 
> Use the export_configuration step to export the configuration to the wanted location.
> 
> Adjust the exported configuration mold for other systems to replicate. For example, remove sections that do not need to be replicated such as iDRAC connection settings. The configuration mold can be accessed directly from the storage location.
> 
> Import the selected configuration mold into the other systems using the import_configuration step.
> 
> It is not mandatory to use export_configuration step to create a configuration mold. Upload the file to a designated storage location without using Ironic if it has been created manually or by other means.

File: docs/source/usage/security_recommendations.rst [hunk 9]
```diff
@@ -112,5 +112,203 @@ Note:
...
+
+  openstack baremetal node clean --clean-steps  '[{"interface": "management","step": "export_configuration","args":{"export_configuration_location": <swift-container-url>}}]' <node>
+  openstack baremetal node clean --clean-steps  '[{"interface": "management","step": "export_configuration","args":{"export_configuration_location": <HTTP-url>}}]' <node>
+  openstack baremetal node clean --clean-steps  '[{"interface": "management","step": "import_export_configuration","args":{"export_configuration_location": <swift-container-url>, "import_configuration_location": <swift-container-url>}}]' <node>
+
+
+Notes:
+  * It is not mandatory to use export_configuration step to create a configuration mold. Upload the file to a designated storage location without using Ironic if it has been created manually or by other means.
+
+  * The HTTP web server does not support multi tenancy and is intended to be used in a stand-alone Ironic, or single-tenant OpenStack environment.
```
[the remark this answers is not in the document]
Comment (mnbatra) [reply docs/source/usage/security_recommendations.rst:307]: Removing this point

File: docs/source/usage/security_recommendations.rst [hunk 10]
```diff
@@ -112,5 +112,203 @@ Note:
...
+     b. **import_configuration** with the **import_configuration_location** input parameter to import the existing configuration mold into the system.
+
+     c. **import_export_configuration** with the **export_configuration_location** and **import_configuration_location input parameters**. This step combines the previous two steps into one step that first imports existing configuration mold into system, then exports the resulting configuration.
+
+
+   * Storage configuration for Swift Containers (Preferred method):
+
+     a. Create the containers to be used for configuration mold storage.
+
+     b. For Ironic Swift user that is configured in the [swift] section add read/write access to these containers.
```
[the remark this answers is not in the document]
Comment (mnbatra) [reply docs/source/usage/security_recommendations.rst:230]: I think I misunderstood this initially, you are right. Swift user will be admins.

File: docs/source/usage/security_recommendations.rst [hunk 11]
```diff
@@ -112,5 +112,203 @@ Note:
...
+     a. To use HTTP server with configuration molds, Enable HTTP PUT support.
+
+     b. Create the directory to be used for the configuration mold storage.
+
+     c. Configure read/write access for HTTP Basic access authentication and provide user credentials in [molds]user and [molds]password fields.
+
+3. Modify or adjust the exported system configuration for other systems to replicate. For example, remove sections that do not need to be replicated such as iDRAC connection settings. The configuration mold can be accessed directly from the storage location.
+
+4. Import the selected configuration mold into the other systems using the import_configuration manual cleaning step.
+
```
[the remark this answers is not in the document]
Comment (mnbatra) [reply docs/source/usage/security_recommendations.rst:243]: Added examples

File: docs/source/usage/security_recommendations.rst [hunk 12]
```diff
@@ -112,5 +112,203 @@ Note:
...
+   * Storage configuration for HTTP servers:
+
+     a. To use HTTP server with configuration molds, Enable HTTP PUT support.
+
+     b. Create the directory to be used for the configuration mold storage.
+
+     c. Configure read/write access for HTTP Basic access authentication and provide user credentials in [molds]user and [molds]password fields.
+
+3. Modify or adjust the exported system configuration for other systems to replicate. For example, remove sections that do not need to be replicated such as iDRAC connection settings. The configuration mold can be accessed directly from the storage location.
```
[the remark this answers is not in the document]
Comment (mnbatra) [reply docs/source/usage/security_recommendations.rst]: Fixed the flow

File: docs/source/usage/security_recommendations.rst [hunk 13]
```diff
@@ -112,5 +112,177 @@ Note:
...
+Sample scripts to export and import the iDRAC configuration using the `Redfish API`_ are shown below. For more information refer the Redfish Module `function definitions`_.
+
+* Export configuration:
+
+.. prompt:: bash $
+
+  python3 ExportSystemConfigurationLocalREDFISH.py -ip '10.1.10.3' -u 'new-user' -p $ADMIN_PASSWORD -t IDRAC
+
+
+* Import configuration after making changes:
```
Comment (tzumainn) [inline docs/source/usage/security_recommendations.rst:166]: The "making changes" here should maybe be its own bullet point before this; after you export the configuration, you make changes, and then you import it. Under the "Make changes" bullet point you can talk a little about reasons why; getting rid of node-specific items or passwords, etc.

File: docs/source/usage/security_recommendations.rst [hunk 14]
```diff
@@ -112,5 +112,177 @@ Note:
...
+      }
+    }
+  }
+
+
+**Steps**:
+
+1. Storage options for Configuration settings:
+
+   * Swift Containers (Admins):
```
Comment (tzumainn) [inline docs/source/usage/security_recommendations.rst:226]: I'd say "Admin-only" and then add an explanation as to why.

File: docs/source/usage/security_recommendations.rst [hunk 15]
```diff
@@ -112,5 +112,177 @@ Note:
...
+
+     a. Create the containers to be used for configuration mold storage.
+
+     b. For Ironic Swift user that is configured in the [swift] section add read/write access to these containers.
+
+
+   * HTTP servers (Owners):
```
Comment (tzumainn) [inline docs/source/usage/security_recommendations.rst:233]: This isn't really "Owners" only; admins can use this approach as well.

File: docs/source/usage/security_recommendations.rst [hunk 16]
```diff
@@ -112,5 +112,177 @@ Note:
...
+     a. To use HTTP server with configuration molds, Enable HTTP PUT support.
+
+     b. Create the directory to be used for the configuration mold storage.
+
+     c. Configure read/write access for HTTP Basic access authentication and provide user credentials in [molds]user and [molds]password fields.
+
+
+2. Use the following manual cleaning steps to export configuration to the wanted location. Run the cleaning step during “manageable” state,  replace the cleaning step and args with one of the following options:
+
+   a. **export_configuration** from the existing system with the **export_configuration_location** input parameter to export the configuration to either a web server(Owners) or a Swift container(Admins).
```
Comment (tzumainn) [inline docs/source/usage/security_recommendations.rst:244]: No need to specify "Owners" or "Admins" here.

File: docs/source/usage/security_recommendations.rst [hunk 17]
```diff
@@ -112,5 +112,177 @@ Note:
...
+
+     openstack baremetal node clean --clean-steps  '[{"interface": "management","step": "export_configuration","args":{"export_configuration_location": <swift-container-url>}}]' <node>
+
+3. Restore the selected system configuration mold using the import_configuration manual cleaning step.:
+
+   a. Directly to the original system with minimal or no modification.
+
+   b. Configure a one to many system configuration settings template i.e.. Modify and import the exported system configuration for other systems to replicate. For example, remove sections that do not need to be replicated such as iDRAC connection settings. The configuration mold can be accessed directly from the storage location.
+
+4. Import Settings either from a web server(Owners) or from a Swift containers(Admins) to the required system:
```
Comment (tzumainn) [inline docs/source/usage/security_recommendations.rst:264]: You can just call this section "Import Settings"

**Review by tzumainn — CHANGES_REQUESTED** [review]
This reads much more cleanly! Some comments in-line, but I think this is close to being merged.

* cleaning.rst probably shouldn't be included here, as that'll be a separate PR (I think?)
* you should squash your commits into one

File: docs/source/usage/security_recommendations.rst (hunk 17 above)
Comment (DanNiESh) [reply docs/source/usage/security_recommendations.rst:264]: Same. No need to specify "Owners" or "Admins".

File: docs/source/usage/security_recommendations.rst (hunk 13 above)
Comment (mnbatra) [reply docs/source/usage/security_recommendations.rst]: Done

File: docs/source/usage/security_recommendations.rst (hunk 14 above)
Comment (mnbatra) [reply docs/source/usage/security_recommendations.rst]: Done

File: docs/source/usage/security_recommendations.rst (hunk 15 above)
Comment (mnbatra) [reply docs/source/usage/security_recommendations.rst]: Removed "Owners"

File: docs/source/usage/security_recommendations.rst (hunk 17 above)
Comment (mnbatra) [reply docs/source/usage/security_recommendations.rst]: done

File: docs/source/usage/security_recommendations.rst [hunk 18]
```diff
@@ -112,5 +112,183 @@ Note:
...
+      * Usage
+           - racadm license help import
+           - racadm license help export
+
+Import or export iDRAC configuration with Ironic manual cleaning :
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The clean and deploy steps provided in this section allow an owner to configure the system and collect the system inventory using configuration mold files. To export or import the configuration of current system or replicate the system configuration to that of a similar system.
+
+The following is an example of a system configuration JSON mold:
```
Comment (tzumainn) [inline docs/source/usage/security_recommendations.rst:189]: I don't think we need this example

File: docs/source/usage/security_recommendations.rst [hunk 19]
```diff
@@ -112,5 +112,183 @@ Note:
...
+
+Configure the node appropriately:
+ The following command configures a bare metal node with the idrac hardware type assuming a mix of Redfish and WSMAN interfaces are used:
+
+.. prompt:: bash $
+
+  openstack baremetal node set \
+    --driver idrac \
+    --driver-info drac_username=user \
+    --driver-info drac_password=pa$$w0rd
```
Comment (tzumainn) [inline docs/source/usage/security_recommendations.rst:141]: missing a '\' at the end here

File: docs/source/usage/security_recommendations.rst [hunk 20]
```diff
@@ -112,5 +112,183 @@ Note:
...
+    --driver-info drac_address=drac.host \
+    --driver-info redfish_username=user \
+    --driver-info redfish_password=pa$$w0rd \
+    --driver-info redfish_address=drac.host \
+    --driver-info redfish_system_id=/redfish/v1/Systems/System.Embedded.1 \
+    --bios-interface idrac-redfish \
+    --inspect-interface idrac-redfish \
+    --management-interface idrac-redfish \
+    --power-interface idrac-redfish
```
Comment (tzumainn) [inline docs/source/usage/security_recommendations.rst:150]: And a missing '\' here

**Review by tzumainn — CHANGES_REQUESTED** [review]
This looks really good! The flow is a lot cleaner now and I think it's much easier for someone to understand the exact details of what they're supposed to do. Three minor comments, and then I think it's good!

**Review by tzumainn — CHANGES_REQUESTED** [review]
This looks really good! The flow is a lot cleaner now and I think it's much easier for someone to understand the exact details of what they're supposed to do. Three minor comments, and then I think it's good!

File: docs/source/usage/security_recommendations.rst (hunk 19 above)
Comment (mnbatra) [reply docs/source/usage/security_recommendations.rst:141]: added

**Comment by mnbatra** [comment]
Ready to merge

**Review by tzumainn — APPROVED** [review]
Great - thanks! This looks really good :)

**Pull request merged**